hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRejectHandler.java

```diff
@@ -118,9 +142,135 @@ public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckExcept
+
+        final var processedRejectionTransfers =
+                processRejectionsForTransferAndRemoveAllowances(rejections, context, rejectingAccountID, hederaConfig);
+
+        final var body = CryptoTransferTransactionBody.newBuilder()
+                .tokenTransfers(processedRejectionTransfers)
+                .build();
+
+        // Apply all changes to the handleContext's States
+        final var transferContext = new TransferContextImpl(context, false, true);
+        final var steps = decomposeTransferIntoSteps(body, context.payer());
+        steps.forEach(step -> step.doIn(transferContext));
+    }
+
+    private List<TokenTransferList> processRejectionsForTransferAndRemoveAllowances(
+            @NonNull final List<TokenReference> rejections,
+            @NonNull final HandleContext context,
+            @NonNull final AccountID rejectingAccountID,
+            @NonNull final HederaConfig hederaConfig) {
+
+        // Initialize stores and get the rejecting account
+        final var tokenStore = context.writableStore(WritableTokenStore.class);
+        final var accountStore = context.writableStore(WritableAccountStore.class);
+        final var nftStore = context.writableStore(WritableNftStore.class);
+        final var relStore = context.writableStore(WritableTokenRelationStore.class);
+
+        final var rejectingAccount =
+                getIfUsable(rejectingAccountID, accountStore, context.expiryValidator(), INVALID_ACCOUNT_ID);
+
+        // Prepare collections for processing
+        final var tokenTransferLists = new ArrayList<TokenTransferList>();
+        final var updatedFungibleTokenAllowances = new ArrayList<>(rejectingAccount.tokenAllowances());
+
+        final var allowancesIsEnabled = hederaConfig.allowancesIsEnabled();
+        for (final var rejection : rejections) {
+            if (rejection.hasFungibleToken()) {
+                processFungibleTokenRejection(
+                        rejection,
+                        rejectingAccountID,
+                        tokenTransferLists,
+                        tokenStore,
+                        relStore,
+                        updatedFungibleTokenAllowances,
+                        allowancesIsEnabled);
+            } else if (rejection.hasNft()) {
+                processNftRejectionAndRemoveSpenderAllowance(
+                        requireNonNull(rejection.nft()),
+                        rejectingAccountID,
+                        tokenTransferLists,
+                        nftStore,
+                        tokenStore,
+                        allowancesIsEnabled);
+            }
+        }
+
+        // Update the account with the new fungible token allowances
+        if (allowancesIsEnabled) {
+            final var updatedAccount = rejectingAccount
+                    .copyBuilder()
+                    .tokenAllowances(updatedFungibleTokenAllowances)
+                    .build();
+            accountStore.put(updatedAccount);
+        }
```

Comment:
I would suggest do this as the end step in handle

Reply:
I've adjusted the logic.